### PR TITLE
Create a model for creating neural networks

### PR DESCRIPTION
**Pull request by antonsakhanovych**
### Notes
1. Created interfaces that neural network will operate on. Main abstraction is Propogatable. Since both Layer and Activation functions count as steps in the computational graph, they are Propogatable. 
2. Since softmax + cross-entropy collapse to `A - Y` to calculate dL/dZ, I decided to couple Loss function with the last layer's activation function. You can see that Softmax is present inside the CrossEntropy class.
3. During backpropagation. Layers get dL/dZ as an input and Activation functions get dL/dA as an input. Only last layer gets dL/dZ directly from the CrossEntropy. Because of that you can see Identity Activation in the main file.
4. Also added Protocols for Updater and Regularizer(Thursday 02/19 lecture). Updater will be responsible for updating the weights. And Regularizer will calculate the penalty for the loss function as well as calculating the penalty term for gradients. 

### Testing
1. Main implements 3layer RandomData homework from the class. The final cost matches.
Note: I am using the technique from early lectures to initialize weights. `torch.sqrt(torch.tensor(2.0 / in_features))`
2. If using the primitive technique, matching the homework, the final cost matches.
snip_20260220_214056
snip_20260220_214135
snip_20260220_214215

[diff.txt](https://github.com/user-attachments/files/25455102/diff.txt)

### CODE REVIEW

File: src/core/protocols.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+class Activation(Propagatable, Protocol):
+    pass
+
+
+class Layer(Propagatable, Protocol):
+    def parameters(self) -> Dict[ParameterType, nn.Parameter]: ...
+
+
+class Loss(Protocol):
```
[the remark this answers is not in the document]
Comment (antonsakhanovych) [reply src/core/protocols.py]: Agree!

File: src/core/activations.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        self._cached_output = output
+        return output
+
+    def backward(self, grad_output: torch.Tensor) -> torch.Tensor:
+        if self._cached_output is None:
+            raise ValueError("Must call forward() before backward()")
+
+        a = self._cached_output
+        grad_input = grad_output * a * (1 - a)
+        return grad_input
```
[the remark this answers is not in the document]
Comment (antonsakhanovych) [reply src/core/activations.py]: Yeah it is the caching from the diagram from 9th lecture notes(9th page).

File: src/core/protocols.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    def backward(self, logits: torch.Tensor, targets: torch.Tensor) -> torch.Tensor: ...
+
+
+class Regularizer(Protocol):
+    def compute_penalty(self, parameters: List[nn.Parameter]) -> torch.Tensor: ...
+    def compute_gradient(self, parameter: nn.Parameter) -> torch.Tensor: ...
+
+
+class Updater(Protocol):
+    def step(self, parameters: Dict[ParameterType, List[nn.Parameter]]) -> None: ...
```
[the remark this answers is not in the document]
Comment (declspecl) [reply src/core/protocols.py]: Ah interesting idea, I see what you mean. I'd say for now we aren't using it, so we can keep it on the backburner or remove it until we have a use case for it

**Review by declspecl — APPROVED** [review]
:lgtm:
